clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+
+  // kernel8 and kernel9 - expect error
+  // Make two lambdas
+  // Call the builtin on the second of the two lambdas (in a constexpr context)
+  // Within a constexpr if, pass the first lambda to a kernel in the true branch,
+  // pass the second lambda to a kernel in the false branch
+  // Test that passing a lambda to the unique stable name builtin and passing it
+  // to a kernel called with an if constexpr branch causes a diagnostic on the
+  // kernel9 use due to the change in the results to the stable name
+  auto l9 = []() { return 1; };
+  auto l10 = []() { return 2; };
+  constexpr const char *l10_output = __builtin_unique_stable_name(l10); // #USN_l10
+  if constexpr (1) {
+    kernel_single_task<class kernel8>(l9);
+  } else {
+    // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+    // expected-note@#USN_l10{{'__builtin_unique_stable_name' evaluated here}}
+    // expected-note@+1{{in instantiation of function template specialization}}
+    kernel_single_task<class kernel9>(l10);
+  }
+
+  // kernel6 - expect no error
+  // Make a lambda
+  // Make another lambda that captures the first one
+  // Call the builtin on the first lambda (in a constexpr context)
+  // Pass the second lambda to a kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda capturing the first one to the kernel does not
+  // throw an error because the first lambda is not involved in naming the
+  // kernel i.e., the mangling does not change
+  auto l5 = []() { return 1; };
+  auto l6 = [l5]() { return 2; };
+  constexpr const char *l5_output = __builtin_unique_stable_name(l5);
+  kernel_single_task<class kernel6>(l6); // #kernel6call
+
+  // kernel11 - expect an error
+  // Make a lambda
+  // Call the builtin on the first lambda (in a constexpr context)
+  // Pass the lambda as a template template parameter to a kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing it to the kernel as a template template parameter causes a
+  // diagnostic on the kernel use due to the change in results to the stable
+  // name
+  auto l11 = []() { return 1; };
+  constexpr const char *l11_output = __builtin_unique_stable_name(l11); // #USN_l11
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_l11{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel11>(S<Tangerine, decltype(l11)>{});
+
+  // kernel12 - expect an error
+  // Test that passing a lambda to the unique stable name builtin within a macro
+  // and then calling the macro within the kernel causes an error on the kernel
+  // due to the change in the results of the stable name
```

Comment:
```suggestion
  // kernel12 - expect an error
  // Tests that passing a lambda to the unique stable name builtin within a macro
  // and then calling the macro within the kernel causes an error on the kernel and
  // diagnoses in all the expected places despite use of a macro.
```

Reply:
Taking all changes except "Tests"; Keeping "Test" to be consistent with all other comments.